awx/main/tasks/jobs.py

```diff
@@ -353,6 +359,52 @@ def create_expect_passwords_data_struct(self, password_prompts, passwords):
             expect_passwords[k] = passwords.get(v, '') or ''
         return expect_passwords
 
+    def release_lock(self, project):
+        try:
+            fcntl.lockf(self.lock_fd, fcntl.LOCK_UN)
+        except IOError as e:
+            logger.error("I/O error({0}) while trying to release lock file [{1}]: {2}".format(e.errno, project.get_lock_file(), e.strerror))
+            os.close(self.lock_fd)
+            raise
+
+        os.close(self.lock_fd)
+        self.lock_fd = None
+
+    def acquire_lock(self, project, unified_job_id=None):
+        if not os.path.exists(settings.PROJECTS_ROOT):
+            os.mkdir(settings.PROJECTS_ROOT)
+
+        lock_path = project.get_lock_file()
+        if lock_path is None:
+            # If from migration or someone blanked local_path for any other reason, recoverable by save
+            project.save()
+            lock_path = project.get_lock_file()
+            if lock_path is None:
+                raise RuntimeError(u'Invalid lock file path')
```

Comment:
Should we do this logic inside get_lock_file method?

Reply:
This isn't really code I'm trying to change, I'm just moving it.

I think a problem with `get_lock_file` raising an exception is that it would preclude the approach here, where it does a `.save()` as a last-ditch fix if it initially returns `None`. On the other hand, I would not want a method named `get_lock_file` to cause a save as a side-effect. So that's my 2-second defense of the structure here, although I'm committed to defending it, I don't have a complete idea of how to restructure it.